### ISSUE TITLE
optimize computing speed of array processing

### ISSUE THREAD

**Issue by prjemian** [sstoupin/dtxrd#2]
> There is one bottleneck with computation of topographs.
> 
> First I read a bunch of .tif or .h5 files (sequential topographs at different positions on the crystal rocking curve). Then I combine all these in 3D array, data[k,j,i] where k index represents the snapshot number (angle on the rocking curve) and [i,j] is the XY location on the topograph. Then I do something straightforward: run a double "for" loop to calculate rocking curve statistics for each [i,j] and this takes time.
> 
> It becomes quite time consuming with ~1000x1000 pixels.
> Is there a way to speed up the python code?

This main loop in topography processing could be optimized for execution time, perhaps by using optimized routines from numpy, scipy, or other numerical analysis packages.  To optimize, we may need to divide this long code block so as to identify routines already implemented in numpy, scipy, or other.

https://github.com/sstoupin/dtxrd/blob/1b2708c813ac59620f0954bb8696c97777525631/rctopo#L448-L658

**Comment by prjemian** [comment]
The comments in the code suggest a set of subdivisions:

* Background determination
* Deglitching procedure
* subtracting bkg since it can vary from pixel to pixel
* define central pixel
* printing and visualization
* PIXEL REJECTION CRITERIA
* extracting curve parameters if the pixel complies with the criterion
* Testing (commented out)

**Comment by prjemian** [comment]
As a first optimization, the printing and visualization should be separated from the computational code.

Background determination and deglitching probably both have library routines that could be used, even using the topo-specific math.

This code, for example, looks as if it may be a weighted average: https://github.com/sstoupin/dtxrd/blob/1b2708c813ac59620f0954bb8696c97777525631/rctopo#L455-L462

**Comment by prjemian** [comment]
@sstoupin emailed:

> Hi Pete,
> 
> Thanks for the quick response and suggestions.
> 
> I saw your edits [here] - makes sense. I will go though LGTM at some point - there are many redundancies for sure.
> 
> The deglitching procedure is something rarely used (its for noisy data) it gets enabled with an option, otherwise not functioning. Needs more testing anyway.
> 
> The bottleneck I think originates from the the double "for" loop where either curvestat or fit1d functions are used (both are my functions, perhaps there are better developments). I think I found now one that could be better. https://lmfit.github.io/lmfit-py/model.html
> 
> this can perhaps replace my fit1d (for gaussian profile fitting)
> 
> Not sure about curvestat - this one finds left/right slope positions (which correspond to FWHM).
> 
> I think some sort of vectorized form for these functions would be ideal, don't know yet if these exist.

**Comment by prjemian** [comment]
I agree that vector functions will replace the need for the double loop.  So far, my code changes are not for speed.  One easy thing to take out of the loop is this part:

```
      if opts.bkg==-1:    
        bkg=0.25*(rcurve[0]+rcurve[1]+rcurve[len(rcurve)-2]+rcurve[len(rcurve)-1]) #; print 'bkg = ', bkg  
        b1=0.5*(rcurve[0]+rcurve[1])
        b2=0.5*(rcurve[len(rcurve)-2]+rcurve[len(rcurve)-1])
        t1=th_beg
        t2=th_end
        A=(b2-b1)/(t2-t1)
        B=b2-A*t2
        bkg_lin=A*th+B
```

At minimum, the `if` decision is executed 10^6 times.  A vector replacement *before* the double `for` loop would execute zero or once depending on `opts.bkg` value.  And, easy to demonstrate.

We should add some timing collection first, so we can identify if any change is an actual improvement of execution time.

**Comment by sstoupin** [comment]
This section is a crude evaluation of the rocking curve background: 
 bkg  - constant, average of 4 points on tails
bkg_lin is a line 
the background can be unique for each pixel but generally very similar (unless it is a bad/hot pixel of area detector). 
What lines in the code would you suggest for time tracking?

**Comment by prjemian** [comment]
For starters, just the overall time, as shown in changeset 2393444
